BUG: #12624 where Panel.fillna() ignores inplace=True

Xbar opened the pull request:
Added if condition for fillna when ndim==3 (for Panel) Issue: https://github.com/pydata/pandas/issues/12624

jreback commented on pandas/core/generic.py:
```diff
@@ -3124,6 +3124,10 @@ def fillna(self, value=None, method=None, axis=None, inplace=False,
+                if inplace:
+                    new_obj = self._constructor.from_dict(result).__finalize__(self)
+                    self._update_inplace(new_obj._data)
+                    return
```
just do an ``else`` here

jreback commented:
this is duped by #12630 but your soln is nicer. Needs tests.

Xbar commented:
Added test case in test_panel.py: test_fillna
Using data provided in issue: https://github.com/pydata/pandas/issues/12624

jreback commented on pandas/tests/test_panel.py:
```diff
@@ -1528,6 +1528,14 @@ def test_fillna(self):
+                      items=['a', 'b'], minor_axis=['x', 'y'])
+        p2 = pd.Panel([[[0, 1], [2, 1]], [[10, 11], [12, 11]]],
+                      items=['a', 'b'], minor_axis=['x', 'y'])
+        p1.fillna(method='ffill', inplace=True)
```
rename ``p2`` -> ``expected``, drop the ``pd.``

jreback replied:
do this in a loop with ``ffill``, ``bfill``

jreback commented:
pls add a whatsnew note as well.

Xbar commented:
Test case changes:
1. Add test case for Panel.fillna, method=‘bfill’
2. Change variable name of expected result to “expected”.
3. Fix discrepancy in dtype by forcing both Panel in this test case to have
dtype=np.float64

Style changes in generic.py:
Added "else" in fillna when ndim==3. Better readability.

jreback commented on pandas/core/generic.py:
```diff
@@ -3124,7 +3124,12 @@ def fillna(self, value=None, method=None, axis=None, inplace=False,
+                if inplace:
+                    new_obj = self._constructor.from_dict(result).__finalize__(self)
+                    self._update_inplace(new_obj._data)
+                    return
```
you don't need this return

jreback commented on pandas/core/generic.py:
```diff
@@ -3124,7 +3124,12 @@ def fillna(self, value=None, method=None, axis=None, inplace=False,
+                    new_obj = self._constructor.from_dict(result).__finalize__(self)
+                    self._update_inplace(new_obj._data)
+                    return
+                else:
```
do something like:

```
result = self._constructor.from_dict(result).__finalize__(self)
if inplace:
    self._update_inplace(result._data)
else:
   return result
```

Xbar replied:
return is necessary, because otherwise it will continue to run to the default code for "2d or less".
But I can refactor this part so that all changes to _data will be put into "new_data" and handled by downstream code.

Xbar replied:
Maybe a silly question, but where should I put the whatsnew note?

jreback replied:
oh i c. ok then. whatsnew in ``doc/source/whatsnew/v0.18.1.txt``

jreback commented:
some small changes. pls add a whatsnew note.

Xbar commented:
1. Updated whatsnew.
2. Style changes in generic.py. 
    -Removed return. 
    -Reduced if branches when ndim==3. 
    -One "if" for inplace at the end of the method. Do not judge inplace twice, unless method is not implemented for some special cases.

jreback commented:
ok looks nice. if you can squash your commits would be great. pls ping on green. (not you had some PEP errors in the prior runs which cause things to fail).
``git diff master | flake8 --diff`` to check/fix.

jreback commented:
https://travis-ci.org/pydata/pandas/jobs/116251314, look at the very end

jreback commented:
pls also rebase on master: ``git rebase -i master`` (this will squash as well)

Xbar commented:
Thanks for the reminder!
Sorry for making so many commits. I'm new to git and version control.

jreback commented:
no, making lots of commits is fine (and good!).

At the end we just like to squash them to a single one.

jreback commented:
this needs to be squashed/rebased on master.

Xbar commented:
I'm not sure if I'm doing it right. I'm running "git rebase -i 648b0f6" and squash all commits I made

jreback commented:
``git rebase -i master``, then once squashed ``git push yourremote thisbranch -f``

jreback commented:
FYI, you don't want to ``git merge master`` normally, instead ``git pull --rebase``. This is the workflow that we use here.

Xbar committed:
BUGFIX: fillNA for panel (ndim==3) Issue: 12624

closes #12624

Xbar commented:
I finally got the idea here. Thanks!
Now it's just 1 commit.

jreback commented:
ok great. ping when green.

jreback commented on doc/source/whatsnew/v0.18.1.txt:
```diff
@@ -89,46 +89,17 @@ Bug Fixes
 - Bug in ``Period`` and ``PeriodIndex`` creation raises ``KeyError`` if ``freq="Minute"`` is specified. Note that "Minute" freq is deprecated in v0.17.0, and recommended to use ``freq="T"`` instead (:issue:`11854`)
-- Bug in ``Series`` construction with ``Categorical`` and ``dtype='category'`` is specified (:issue:`12574`)
-
-
```
in the future, don't take blank lines the whatsnew, just make your changes. I put these in because then we don't have conflicts with multiple PR's.

Xbar replied:
I see. git is telling me conflicts in whatsnew and I got confused.

Xbar commented:
All checks passed, ready to merge.

jreback commented:
thanks @Xbar
